LDAP fixes based on search-by-email.

mateuszmandera opened the pull request:
This should fix the following issues:
https://github.com/zulip/zulip/issues/11878
https://github.com/zulip/zulip/issues/11715
https://github.com/zulip/zulip/issues/9277
possibly some other related ones :thinking: 

1. We swtich to our fork of django-auth-ldap, where I implemented searching in LDAP by email.
2. . A rather unpleasant migration of tests to slapdtest. This is where most of the large number of line changes of PR comes from. This is necessary, because as explained the main migration commit:
> the fakeldap setup we used, with mocking of django_auth_ldap.config.ldap.initialize, would permanently affect the internal state of the ldap modules for all other tests

so having any fakeldap-based tests broke slapdtest (the ldap modules would keep trying to connect to fakeldap instead of slapd.). Plus switching to slapdtest has various other benefits, so I think we're definitely better off relying on that everywhere.

3. We fix email_belongs_to_ldap, to search for emails in the directory rather than blindly returning True.

4. Use email search in django_to_ldap_username, which allows fixing the flows of things like getting api keys and sync_user_from_ldap, which use user_profile.email as the identifier, by allowing them to translate that email into the proper ldap username.

timabbott commented:
When trying to run a single test class from this with `time test-backend LDAPUserSignUpTest`, I get this failure:

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/suite.py", line 163, in _handleClassSetUp
    setUpClass()
  File "/home/tabbott/zulip/zerver/lib/test_classes.py", line 899, in setUpClass
    cls.server.start()
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/slapdtest/_slapdtest.py", line 445, in start
    self._start_slapd()
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/slapdtest/_slapdtest.py", line 420, in _start_slapd
    raise RuntimeError("slapd exited before opening port")
RuntimeError: slapd exited before opening port

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tabbott/zulip/tools/test-backend", line 473, in <module>
    main()
  File "/home/tabbott/zulip/tools/test-backend", line 391, in main
    include_webhooks=include_webhooks)
  File "/home/tabbott/zulip/zerver/lib/test_runner.py", line 555, in run_tests
    result = self.run_suite(suite)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/test/runner.py", line 567, in run_suite
    return runner.run(suite)
  File "/usr/lib/python3.6/unittest/runner.py", line 176, in run
    test(result)
  File "/usr/lib/python3.6/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/home/tabbott/zulip/zerver/lib/test_runner.py", line 350, in run
    self._handleClassSetUp(test, result)  # type: ignore
  File "/usr/lib/python3.6/unittest/suite.py", line 170, in _handleClassSetUp
    self._addClassOrModuleLevelException(result, e, errorName)
  File "/usr/lib/python3.6/unittest/suite.py", line 216, in _addClassOrModuleLevelException
    result.addError(error, sys.exc_info())
  File "/home/tabbott/zulip/zerver/lib/test_runner.py", line 172, in addError
    test_name = full_test_name(args[0])
  File "/home/tabbott/zulip/zerver/lib/test_runner.py", line 70, in full_test_name
    test_method = test._testMethodName
AttributeError: '_ErrorHolder' object has no attribute '_testMethodName'
```

Any idea what the issue is?

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +856,85 @@ def setUp(self) -> None:
+        cls.server.start()
+
+        ldif = cls.fixture_data('directory.ldif', type='ldap')
+        cls.server.ldapadd(ldif)
```
The thing I don't like with `slapdtest` in comparison to the previous `fakeldap` system is the need to run a service, which can introduce reliability issues (like the traceback above).  

How should one the `.ldif` extension?

mateuszmandera replied:
@timabbott Do you think it would be preferrable to try expand fakeldap's search_s to work for our email-search testing purposes, rather than stick with slapdtest?

> How should one the .ldif extension?

I think the verb got lost in typing here? :thinking:

timabbott replied:
Yes, I meant "how should one read the `.ldif` extension?"

timabbott replied:
Oh, and for your actual question: I think if we can expand `fakeldap` to support this kind of query, that'd be a totally reasonable solution.  We do control the `fakeldap` library, so we can merge changes to expand it if we need to.

mateuszmandera replied:
N/A anymore, since things are switched to a fakeldap setup

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +856,85 @@ def setUp(self) -> None:
+            "core.schema",
+            "cosine.schema",
+            "inetorgperson.schema",
+        ]
```
Can you add a comment explaining this list of schemas?  E.g. are they the standard base schemas present in any LDAP server, or just ones used for this particular purpose?

mateuszmandera replied:
N/A anymore, since things are switched to a fakeldap setup

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +856,85 @@ def setUp(self) -> None:
+        super().tearDownClass()
+        cls.server.stop()
+
+    def settings(self, **kwargs: Any) -> Any:
```
Let's do a comment at the top of this function:
"Extend `self.settings` for LDAP tests to set various LDAP defaults if not explicitly specified: The server URI, and how to find users by username or email".

mateuszmandera replied:
Added

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +856,85 @@ def setUp(self) -> None:
+                                                         ldap.SCOPE_SUBTREE, "(uid=%(user)s)")
+        if not kwargs.get("LDAP_APPEND_DOMAIN"):
+            # If LDAP_APPEND_DOMAIN is not used, email search should be configured.
+            # Use defaults here, unless the test specifies its own settings for these.
```
Let's explain this more clearly: "If LDAP_APPEND_DOMAIN is specified, then AUTH_LDAP_USER_SEARCH can be used to look up users by email address.  Otherwise, we need to declare how to look up users by email address."

mateuszmandera replied:
Added

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +856,85 @@ def setUp(self) -> None:
+        return super().settings(**settings_kwargs)
+
+    def change_user_attr(self, username: str, attr_name: str, attr_value: str,
+                         binary: bool=False) -> None:
```
We should add a docstring for this explaining that it's used to modify LDAP attributes during the runtime of a test.

We should also explain clearly the duration of these changes -- does it last just for the duration of the current test function, or the whole test class?

mateuszmandera replied:
Added, though the matter is much more clear now that things are back on fakeldap - each test function has its own version of the directory.

timabbott commented on zerver/tests/fixtures/ldap/directory.ldif:
```diff
@@ -0,0 +1,89 @@
+dn: dc=zulip,dc=com
```
Let's add a comment at the top explaining that this is the test LDAP database used in the development environment.

mateuszmandera commented:
> When trying to run a single test class from this with ``time test-backend LDAPUserSignUpTest, I get this failure``

Hmm, looking at slapdtest code it seems like a generic error that gets thrown if slapd doesn't start:
```
        self._proc = subprocess.Popen(slapd_args)
        # Waits until the LDAP server socket is open, or slapd crashed
        # no cover to avoid spurious coverage changes, see
        # https://github.com/python-ldap/python-ldap/issues/127
        for _ in range(10):  # pragma: no cover
            if self._proc.poll() is not None:
                self._stopped()
                raise RuntimeError("slapd exited before opening port")
```

is that only for this specific command, or running any of these ldap tests with tools/test-backend ?

timabbott commented on zerver/tests/test_ldap.py:
```diff
@@ -0,0 +1,109 @@
+    def test_authenticate_to_ldap_via_email(self) -> None:
+        """
+        With django_to_ldap_username now able to translate an email to ldap username,
+        it should be possible to authenticate through user_profile.email.
```
Code docstrings shouldn't say things like "now able to translate ...", because after these changes are merged, the historical fact that this feature was added later won't be relevant to anyone.  We should instead talk about "With AUTH_LDAP_REVERSE_EMAIL_SEARCH configured, ...`

mateuszmandera replied:
Changed to
```
        """
        With AUTH_LDAP_REVERSE_EMAIL_SEARCH configured, django_to_ldap_username
        should be able to translate an email to ldap username,
        and thus it should be possible to authenticate through user_profile.email.
        """
```

timabbott commented:
I merged the first commit (after changing the fake version number to `2.0.0zulip1` and updating `PROVISION_VERSION`) and posted a bunch of comments.

mateuszmandera commented:
Oh, could that have been a slapd crash due to incorrect provisioning, because I forgot to bump PROVISION_VERSION?

timabbott commented:
> Oh, could that have been a slapd crash due to incorrect provisioning, because I forgot to bump PROVISION_VERSION?

Well, I did run provision first; all `PROVISION_VERSION` does is determine whether to force one to run provision.  And I think it's starting a separate copy of slapd for this purpose.  So I think we're missing something here.

mateuszmandera commented:
@timabbott I pushed with changes, with the tests now based on fakeldap again - but I believe a lot of the test migration I did for slapdtest still useful, so it's mostly still there - now the ldap tests use ``LDAPTestCase`` as a base class, similar in spirit to the slapdtestcase class from before.

There is a commit switching requirements to use fakeldap from my repository, for the sake of making sure circleci passes - it shouldn't be merged, and instead changed accordingly when https://github.com/zulip/fakeldap/pull/14 gets merged.

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +854,68 @@ def setUp(self) -> None:
+class LDAPTestCase(ZulipTestCase):
+    """
+    Base class for LDAP tests. Takes care of the mock_ldap setup, loads
+    a directory from fixtures/ldap/directory.json with various entries to be
```
Let's use the full path to `directory.json` from the root of the project, to make sure it's easy to find, since that's a file folks will often want to reference to understand a given test.

mateuszmandera replied:
Since the logic that I kept from this class is now in methods added to ZulipTestCase instead, the analogy of this docstring is not on the init_default_ldap_database now, with the directory path fixed.

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +854,68 @@ def setUp(self) -> None:
+        """
+        if "AUTH_LDAP_USER_SEARCH" not in kwargs:
+            kwargs["AUTH_LDAP_USER_SEARCH"] = LDAPSearch("ou=users,dc=zulip,dc=com",
+                                                         ldap.SCOPE_ONELEVEL, "(uid=%(user)s)")
```
I think this patching process is somewhat dangerous, because it affects all `with self.settings()` calls in the test case, not just those that mean to be affecting LDAP settings.  I think it might be cleaner to just use a special value, e.g. `AUTH_LDAP_USER_SEARCH=self.default_ldap_search_config()`, rather than having this hack.

timabbott replied:
(I think this lets us remove this overridden function entirely)

mateuszmandera replied:
Instead of overriding the settings() method like this, I figured that it seems most sensible to just put the configuration in ``test_settings.py``, the straight-forward way. I hope that's a more reasonable approach - I don't think having those settings in all tests affects anything important - and the tests that don't want them, or want different ones, should override them explicitly in my opinion.

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +854,68 @@ def setUp(self) -> None:
+            settings_kwargs[key] = value
+        return super().settings(**settings_kwargs)
+
+    def change_user_attr(self, username: str, attr_name: str, attr_value: Union[str, bytes],
```
Let's call this `change_ldap_user_attr`.

mateuszmandera replied:
Changed (this is now a method of ZulipTestCase).

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -850,3 +854,68 @@ def setUp(self) -> None:
+
+    def tearDown(self) -> None:
+        self.mock_ldap.reset()
+        self.mock_initialize.stop()
```
If we wanted to try not having a separate test class, I think we could do `if self.mock_ldap is not None:` for these, and then just have tests intending to use LDAP replace the `setUp` method with the test calling `self.initialize_default_ldap_database()` at the start.  

(This potentially should replace the code in ZulipLDAPAuthBackendBase that conditionally calls `init_fakeldap`, which was I think intended to do what you're trying to do here).

mateuszmandera replied:
I added this at the end of ZulipTestCase's tearDown. Seems like a good follow up indeed to unify things with init_fakeldap, I'll try to get that done next.

mateuszmandera commented:
@timabbott  I changed the tests approach, to be similar, but based on ZulipTestCase now

timabbott commented:
I noticed your fixing a bug where we didn't call the `super()` method for `setUp` in one of our test classes; we may want to do a quick audit to check we don't have that mistake in setUp/tearDown in other test classes.  Could have nasty side effects.

timabbott commented:
OK, I fixed one merge issue with d364891894d4dc5f87c811293644adaca9817d9d and then merged the first N commits.  A few comments for the remaining commits:

* `email_belongs_to_ldap` should probably use an early-return code path (i.e. flipping around the conditional for `LDAP_APPEND_DOMAIN` to be less nested, and similarly for the other settings check).
* This comment: `# aaron's uid in our test directory is "letham".` might be better replaced with a `self.ldap_username(aaron)` sort of thing (see comment below about passwords below), plus a comment about how username is unrelated to his email.
* Does `ldap: Do a proper search for email in email_belongs_to_ldap.` close any issues?  I would also start with describing the bug in the commit message.

A few things I notice about the tests that we should probably fix as follow-ups:
* Let's change the default password for hamlet/newuser/etc. in LDAP to e.g. `hamletPassword` (and similarly for the other LDAP users); the for some of these tests is much less readable with e.g. `testing` as the password when we're not setting that value earlier in the test case.  Or maybe better yet, `self.ldap_password("hamlet")` so it's completely clear that this is the user's actual LDAP password.  Will let us remove some silly comments as well.
* The thing I noted about wanting to do a `super()` audit in tearDown/setUp methods
* The commit adding `AUTH_LDAP_REVERSE_EMAIL_SEARCH` should add content to the ReadTheDocs docs as well.

Also, for `ldap_to_django_username`, can you explain the reasoning behind the choice to not do the mapping to Django username there?  I don't understand why having a separate `user_email_from_ldapuser` function that callers access is the right model.

mateuszmandera commented:
> Also, for ldap_to_django_username, can you explain the reasoning behind the choice to not do the mapping to Django username there? I don't understand why having a separate user_email_from_ldapuser function that callers access is the right model.

Because if we're using LDAP_EMAIL_ATTR, then the django username (email) can only be obtained from the _LDAPUser object. ``ldap_to_django_username`` takes only username (string) as argument, so it's not capable of doing this logic. 

As you can see here https://github.com/zulip/zulip/pull/13256/commits/3151ef612e77c287d9be97d1362eaa903a9135dd#diff-b30c5f6379655d0b194681a8dd57e9a7L544 we already were doing this logic in ZulipLDAPAuthBackend.get_or_build_user, but it's also needed in ZulipLDAPUserPopulator.get_or_build_user, so it seems best to have one function to call from those places.

mateuszmandera commented on zerver/tests/test_ldap.py:
```diff
@@ -0,0 +1,91 @@
+
+class DjangoToLDAPUsernameTests(ZulipTestCase):
+    def setUp(self) -> None:
+        super().setUp()
```
This super().setUp() is actually useless, it's a leftover from when things were based on the LDAPTestCase class, which did have its own setup() that needed to be called by subclasses. I'll get rid of this here.

mateuszmandera replied:
But this whole thing made me notice a more important thing - tearDown(). We actually do have some tearDown() code in ZulipTestCase and I think we should make sure that any tests that override tearDown remember to run super().tearDown() - made a PR with more elaboration at https://github.com/zulip/zulip/pull/13300 @timabbott

mateuszmandera commented:
>The commit adding AUTH_LDAP_REVERSE_EMAIL_SEARCH should add content to the ReadTheDocs docs as well.

Fixed in the ``ldap: Do a proper search for email in email_belongs_to_ldap. `` commit. 

>Does ldap: Do a proper search for email in email_belongs_to_ldap. close any issues? I would also start with describing the bug in the commit message.

Should be fixing https://github.com/zulip/zulip/issues/11715 - I added "Closes #11715" in the commit message. The rest of the message already has some elaboration, should I improve it somehow?

timabbott commented:
> Because if we're using LDAP_EMAIL_ATTR, then the django username (email) can only be obtained from the _LDAPUser object. ldap_to_django_username takes only username (string) as argument, so it's not capable of doing this logic.

OK, cool, that makes sense.  We might want to open an issue with upstream (or a comment on our PR for the overall LDAP_EMAIL_ATTR feature) requesting an alternative API to `ldap_to_django_username` that is passed a `_LDAPUser` object (the LDAP backend could check if this `ldap_user_to_django_username` function exists before deciding what to pass), just to note that would be the better way to structure this interface.

mateuszmandera commented:
I made a PR for this in their repository https://github.com/django-auth-ldap/django-auth-ldap/pull/154

mateuszmandera commented:
@timabbott Looking at the activity there, it seems upstream hasn't been very active lately, so I don't think we can wait to see what they do with it. We could merge it into our repository, but the more we diverge in it from upstream, the more tricky it could end up being to mantain it in the future, if they decide they don't want these changes :thinking:

timabbott commented:
Yeah, I think this is one where it just lets us save a bit of code (vs. making the difference in whether it's possible to do what we're trying to do or not).  So I think we should keep our code as-is and we can refactor if/when upstream responds to that PR.

I agree they seem inactive right now, but they've been good maintainers in the past, so we can certainly hope they'll become active again at some point.  If they don't, we can try pinging the maintainer about adopting the package.

timabbott commented:
I did some edits to the docs in the first commit that should hopefully simplify them, and pushed back to this PR.  A few requests for the following commits:
* The tests in the second commit with `find_ldap_users_by_email` should include what happens in cases where there are no results; it's not clear to me whether that function will actually return something falsey when there are no matches.  (I think it returns `[]`, in which case we should be checking its length to make the type clear?)
* I'm not sure we need a separate `test_ldap.py` vs. putting those tests in a new test class in `test_auth_backends.py`; what's the thinking there?
* There are a lot of new if statements in `django_to_ldap_username` where I'm not sure our error handling behavior is correct; e.g. `len(results)` is 0, we fall through to returning the username; is that the right thing to do?  Or should we throw an exception?
* I think we should be moving to a model where `settings.AUTH_LDAP_USERNAME_ATTR` and `settings.AUTH_LDAP_REVERSE_EMAIL_SEARCH` are required if `LDAP_APPEND_DOMAIN` isn't set, and being more forceful about raising an exception if those don't return results for a user.  

A few notes on follow-ups:
* We should update `manage.py query_ldap` to have a appropriate test tooling for the email -> username lookups.  I think it can probably call the reverse mapping logic and check that it works adn returns the original username?
*

mateuszmandera commented:
> The tests in the second commit with find_ldap_users_by_email should include what happens in cases where there are no results; it's not clear to me whether that function will actually return something falsey when there are no matches. (I think it returns [], in which case we should be checking its length to make the type clear?)

This is already, tested in ``test_django_to_ldap_username_with_email_search``, I added a comment above the relevant test line to make it clearer:
```
            # If there are no matches through the email search, return the email unchanged:
            self.assertEqual(self.backend.django_to_ldap_username("no_such_email@example.com"),
                             "no_such_email@example.com")
```

But find_ldap_users_by_email actually returns None if nothing is found, so I fixed the return type in the first commit, and added a docstring to explain the return values explicitly.
```
def find_ldap_users_by_email(email: str) -> Optional[List[_LDAPUser]]:
    """
    Returns list of _LDAPUsers matching the email search,
    or None if no matches are found.
    """
```

mateuszmandera commented:
> There are a lot of new if statements in django_to_ldap_username where I'm not sure our error handling behavior is correct; e.g. len(results) is 0, we fall through to returning the username; is that the right thing to do? Or should we throw an exception?

My reasoning when making it behave this way was that for example we could be dealing with an unlikely case in configuration (C) where the usernames look like emails, but don't match the actual email addresses in ``LDAP_EMAIL_ATTR attribute. For example
```
dn: uid=hamlet@corp_A.com,ou=users,dc=example,dc=com
uid: hamlet@corp_A.com
mail: hamlet@actualcontactaddress.com
```

Then ``django_to_ldap_username(hamlet@corp_A.com)`` won't find results through the email search, and the actual correct username that we want it to return is ``hamlet@corp_A.com``.

Kind of a weird case, but I figured it's better to have that covered.

mateuszmandera commented:
> We should update manage.py query_ldap to have a appropriate test tooling for the email -> username lookups. I think it can probably call the reverse mapping logic and check that it works adn returns the original username?

This should already work, because query_ldap uses ``django_to_ldap_username``, so that should successfully translate email to the user entry. I pushed a commit changing the query_ldap tests a little bit, and one of them now actually verifies this.

timabbott commented:
OK, that reasoning makes sense.  Certainly we're at a point where merging this and fixing forward will be a better experience.  As a small note, it looks like you neglected to fetch my last batch of changes to the settings.py/docs in the first commit, so I restored those from my reflog before merging.

I did a bit more work on the docs/comments and merged this as the series ending with 869b57b7f783131ee56054bb3cd1ab96aca26bbe.  There are a few things we should fix forward:
* We should rework the `self.ldap_password` function to accept the username as an argument and give the users different passwords.  Having every LDAP user possess the same password is a bug land mine where the test isn't testing what one thinks it is.
* I think it's OK to have a separate file for some LDAP testing, but my view on the terminology is that the LDAP code is all within the scope of our LDAP authentication backend, even if it supports syncing data as well as authentication.  It's certainly a Django authentication backend class.  I agree `test_auth_backends` is kinda giant; I'm not sure this is the best way to split it up though.  
* We should look at tweaking the `query_ldap` code to give a good error message if `backend.django_to_ldap_username` fails, since that's now a likely failure mode with broken configuration, and I suspect the errors will be confusing with the current code, since e.g. `find_ldap_users_by_email` with fail and we'll just fail to convert the username.  Probably `query_ldap` should pass an option that gets passed through to `django_to_ldap_username` here that requires a match and gives a nice error if there isn't one?  Not sure.

Huge thanks for fixing this @mateuszmandera!
